docs/src/design/decoder/constraints.md

```diff
@@ -127,30 +129,30 @@ $$
 h_{res} = \alpha_0 + \alpha_1 \cdot m_{hout} + \alpha_2 \cdot (a + 7) + \sum_{i=0}^3(\alpha_{i + 8} \cdot h_i)
 $$
 
-Using the above variables, we define operation values as described below.
+Using the above variables, we define operation values as described below.  It should be noted that the $\alpha_5$ term is multiplied by the [opcode value](../stack/op_constraints.md#high-degree-operations) of the respective control block initializer.
 
-When `JOIN` operation is executed, a new hasher is initialized and contents of $h_0, ..., h_7$ are absorbed into the hasher. We also indicate that the number of elements to be hashed is $8$:
+When `JOIN` operation is executed, a new hasher is initialized and contents of $h_0, ..., h_7$ are absorbed into the hasher. The number of elements to be hashed is $8$ so the $\alpha_4$ term is set to $0$:
 
 $$
-u_{join} = f_{join} \cdot (h_{init8} + \alpha_4 \cdot 8) \text{ | degree} = 7
+u_{join} = f_{join} \cdot (h_{init8} + \alpha_5 \cdot 90) \text{ | degree} = 7
 $$
```

Comment:
Looking at these constraints now, I think we can simplify them quite a bit. Specifically, we can combine constraints fro `JOIN`, `SPLIT`, `LOOP`, `CALL, and `SYSCALL` into a single expression.

First, we can define domain value $d$ in the way you had it defined before (based on `op_bits` registers).

Then, we can have a composite flag $f_{ctrli} =f_{join} + f_{split} + f_{loop} + f_{call} + f_{syscall}$ (this is probably similar to the `ctrlb` flag you had before).

Then, instead of having individual expressions for these operations, we can define a single expression like so:

$$
u_{ctrli} = f_{ctrli} \cdot (h_{init} + \alpha_5 \cdot d)) \text{ | degree} = 7
$$

And then the final constraint becomes:

$$
p_0' \cdot (u_{ctrli} + u_{span} + u_{respan} + u_{end} + 1 - (f_{ctrli} + f_{span} + f_{respan} + f_{end})) = p_0
$$

Reply:
updated